### ISSUE TITLE
Why CodeCrunch does not allow us to declare final static variable

### ISSUE THREAD

**Issue by zzkzzzz** [nus-cs2030/2021-s1#112]
During lab and project, I would like to declare some constant variables to `final static`.
But CodeCrunch will return `*** not declared private final. Grading terminated.`

Why CodeCrunch does not allow us to declare `final static` variable?

**Comment by domlimm** [comment]
Hi @zzkzzzz , I just changed my SERVICE_TIME in the project to `private final static double SERIVCE_TIME = 1.0` and CodeCrunch accepted it. Weird...

Could you maybe show how you did it? Like the declaration and usage of the variable

**Comment by marcuspeh** [comment]
It's just that code crunch is checking for all the variable to be private final. Cuz theres a static in between, it wont pass the check.

@domsterthebot You could do that but it will cause a warning on checkstyle. My TA told me to follow checkstyle

**Comment by domlimm** [comment]
> It's just that code crunch is checking for all the variable to be private final. Cuz theres a static in between, it wont pass the check.
> 
> @domsterthebot You could do that but it will cause a warning on checkstyle. My TA told me to follow checkstyle

Ahh dang it. Thanks marcus! Yeah just checked and my checkstyle isn't clean no more...

**Comment by HolmesJJ** [comment]
Yes I also met the same problem. When I write private final static, the code cannot pass the checkstyle. When I change to private static final, then codecrunch return error... LOL.

**Comment by eman-kom** [comment]
so static is not allowed?

**Comment by SkyBlaise99** [comment]
i guess not, you can just hard code the 1 in lol. I think we might use something similar to SimulationParameter.java in Mock PE in the future.

**Comment by geraldchy** [comment]
I'm having the same problem too, codecrunch forces all variables to be private final so if I have a static variable that stores the num of served customers, my code won't work anymore since private final static variables cannot be modified. Any solutions to this problem?

**Comment by tlylt** [comment]
> I'm having the same problem too, codecrunch forces all variables to be private final so if I have a static variable that stores the num of served customers, my code won't work anymore since private final static variables cannot be modified. Any solutions to this problem?

@Gernnon Maybe go for an approach that doesn't use private final static variables, it is possible to do without it.

**Comment by geraldchy** [comment]
> > I'm having the same problem too, codecrunch forces all variables to be private final so if I have a static variable that stores the num of served customers, my code won't work anymore since private final static variables cannot be modified. Any solutions to this problem?
> 
> @Gernnon Maybe go for an approach that doesn't use private final static variables, it is possible to do without it.

I know it is possible but it just doesn't make sense for it not to be a global class variable hence static.

**Comment by Henry-Yuan-He** [comment]
Since prof wants us to write the immutable class, every attribute in the class must be "private final". So it is impossible for us to use static as an tool to help you automatically add id when creating a new instance.
